### ISSUE TITLE
DataParallel does not work with tensors of dimension 0

### ISSUE THREAD

**Issue by 0phoff** [pytorch/pytorch#9811]
## Issue description
I have a network that return a single value, which is a dimensionless tensor as of PyTorch 0.4.0
However, when I wrap my model in a ``torch.nn.DataParallel`` for execution on multiple GPUs, it complains about not being able to gather the different tensors because they have no dimension.  
imo the tensors should then be combined in a new dimension?  

My specific usecase, is that I have my loss functions in the models, and I managed to fix it by setting ``reduce=False`` in the CrossEntropyLoss. This means I do get a Tensor with a dimension, and thus DataParallel can in fact gather the different tensors. All I have to do is compute the mean of the different tensors myself.

## Code example
```python
# Untested code, but this is the basically what my model does at the end
class Model(nn.module):
  def __init__(self):
    self.loss = torch.nn.CrossEntropyLoss()
  def forward(self, data, target):
    return self.loss(data, target)

m = torch.nn.DataParallel(Model()).cuda()

data = ...    # get tensor for data
target = ...  # get tensor for target

m(data, target)
```

## System Info
Pytorch version : 0.4.0a0
python 3.6

**Comment by sharpsy** [comment]
I believe that it is not a good idea for PyTorch to combine those return values in a new dimension. This change would make a shape of the return value dependent on the number of GPU devices used to calculate the result. As I see the `DataParallel` class, its job is to transparently split the workload on available GPUs so the result is indistinguishable from the result you would get from running the model without the `DataParallel`.

I have not used models that calculate the loss in the forward pass so I do not have an experience for a valid advice - to me it looks like something to avoid. What you can do to implement it with `DataParallel` is to include 'DataParallel' as part of your model. For example, look at this code snippet for the general idea: https://pytorch.org/tutorials/beginner/former_torchies/parallelism_tutorial.html#dataparallel. You could do the same, have all layers in (a single) `DataParallel` and then calculate the loss after this `DataParallel` finishes.

**Comment by 0phoff** [comment]
Yeah but then the loss computations are not split over the different GPUs, thus increasing the training time.

This problem is also bigger than computing losses in the forward pass of a module. Any network that outputs only one number will have this problem with the `DataParallel` class.

Your point about the output being dependent on the number of GPUs is valid though. Not sure how to tackle this in a clean way...
Maybe the `DataParallel` class should have `reduce` and `size_average` parameters like the different loss functions for aggregating the output. 

Setting `reduce` to __False__ would result in the current behaviour.  
Setting it to __True__ would mean we take the sum/average of the different output tensors (depending on `size_average`).  
We could even allow the creation of a new dimension if `reduce` is __False__  and the outputs a scalar tensors. I know this means the output is dependent on the number of GPUs but that would mean the user would want that in that case?

**Comment by sharpsy** [comment]
It is true, the loss computation could increase the training time. My hunch is that loss calculation is usually the fast part of the model training. I'm thinking of all other computation that takes place besides the loss calculation - forward pass, gradient calculation, weight updates...

To check this, I just benchmarked `CrossEntropy` on 10000 items and 10000 classes and loss computation took 2.16ms. I used the following benchmark:
`python3 -m timeit -s 'import torch as t; dev="cuda"; loss=t.nn.CrossEntropyLoss().to(dev); inp=t.randn(10000, 10000).to(dev); tgt=t.empty(10000, dtype=t.long).random_(10000).to(dev)' 'loss(inp, tgt)'`

It would seem that this is unlikely to observably influence the training time.

Even if it does, implementation that distributes the loss computation over the GPUs and then calculates the mean of the returned vector is considerably slower.
Try it:
`python3 -m timeit -s 'import torch as t; dev="cuda"; loss=t.nn.DataParallel(t.nn.CrossEntropyLoss(reduce=False)).to(dev); inp=t.randn(10000, 10000, requires_grad=True).to(dev); tgt=t.empty(10000, dtype=t.long).random_(10000).to(dev)' 't.mean(loss(inp, tgt))'`

Regardless of the performance impact, the `DataParallel` has a nice property that the `model(input)` will be equal to `DataParallel(model)(input)`. The proposed change would break this property.

**Comment by SsnL** [comment]
Fixed on master

**Comment by SsnL** [comment]
Allowing modules that return scalar to be used in DataParallel (will return a vector).

**Comment by sharpsy** [comment]
@0phoff , just to add an idea to your use case that minimizes the amount of data transfered between GPUs. You can use `size_average=False, reduce=True` as arguments. Losses on each GPU will be summed, but not divided by the batch-size-on-GPU. You then sum all parallel losses and divide by the size of the whole batch.

**Comment by apaszke** [comment]
I'm also not very happy with stacking the scalars. We've discussed that in #7973, where he claimed that this was the behavior before his fix, so I passed on arguing further. However, if this wasn't the case in the previous release, then I do think we should raise an error if we see scalar outputs. @SsnL?

**Comment by SsnL** [comment]
I agree that returning a vector is unexpected here. But I was under the
impression that 0.3.1 (before the tensor variable merge) behaved that way.
I’ll test and report back here.

**Comment by SsnL** [comment]
@apaszke @0phoff Here is what it used to be 
```py
>>> import torch
>>> torch.__version__
'0.3.1.post3'
>>> import torch.nn as nn
>>> import torch.nn.functional as F
>>>
>>> class Model(nn.Module):
...     def forward(self, x):
...         return F.mse_loss(x, torch.autograd.Variable(torch.ones(x.shape).cuda()))
...
>>> m = nn.DataParallel(Model(), [0, 1])
>>> print(m(torch.autograd.Variable(torch.zeros(4, 3))))
Variable containing:
 1
 1
[torch.cuda.FloatTensor of size 2 (GPU 0)]

```

I didn't really add any new behavior.

**Comment by 0phoff** [comment]
@SsnL, I guess pytorch 0.3.1 had this unintended behaviour because there were no scalar tensors.  
While you indeed set the behaviour back to how it was, I am not sure this is what we want for PyTorch?
This would mean that wrapping a model in `DataParallel` is not a transparent behaviour as __sharpsy__ explained.

Anyway, I am not a big contributor, so I'll let you guys figure this one out :innocent:

**Comment by sharpsy** [comment]
Current behavior leads to fickle code. If backwards compatibility is the reason to leave it as it is - version 1.0 is a great time to solve this.
In https://github.com/pytorch/pytorch/pull/7973, @SsnL wrote down other reasons for this behavior.

> If people want to calculate their losses at each GPU, they can't treat DP as non-existent and will have to face this problem no matter what.

Parallel loss calculation is a valuable feature, and we want to keep it. As the example I wrote showed, automatic loss function reduction in the `DataParallel` can easily lead to wrong results. The (only) way for it to work is with argument `average=False` and then user sums all losses and divide it with the whole batch size. The problem here is that broadcasting scalars to a tensor will be done implicitly and the user needs to combine them explicitly. To do it correctly, user will need to understand what happened in the `DataParallel` implementation. If he understands `DataParallel` implementation, it should not be a problem to come up with `return scalar.unsqueeze(0)` to have current behavior. If he doesn't understand - he will probably have a bug in the code.

> Even if people use .unsqueeze(0), the output shape will still be number_of_GPU-dependent.

Yes, but this is the point - for them to request this behavior explicitly. In that case, we should assume that they know what they are doing.

> This pattern works before 0.4.

The backwards compatibility is a reason to keep this behavior. If it is the only reason, version 1.0 is a great time to deprecate it.

> The pattern @colesbury and I were thinking about is calculating loss with size_average=False in Module forward, and sum on the resulting vector from DP, and backward.

This is a valid pattern and it should stay possible. And it will be possible even if scalars are not broadcasted to tensor (ie. with the `.unsqueeze(0)` and similar tricks).

If outright removal of this 'feature' is currently not an option, I would suggest deprecating it as soon as possible.

**Comment by SsnL** [comment]
I'm fine with a deprecation warning. :)

**Comment by oroojlooy** [comment]
Is there any update on this problem? I still face same problem in version 1.0.

**Comment by cbasavaraj** [comment]
Me too!

**Comment by Dwrety** [comment]
There might be a hack around this. You can specify the batch size dynamically by setting the batch size per gpu. This way you know that the batch size will always be divisible across the number of gpus you are using. And then simply call scalar.mean().

**Comment by iamlockelightning** [comment]
in pytorch 1.8. same warning.
